qos/cosmos/response.go

```diff
@@ -72,7 +73,14 @@ func unmarshalResponse(
 	}
 
 	// Validate the JSON-RPC response.
-	if err := jsonrpcResponse.Validate(getExpectedResponseID(jsonrpcResponse, isJSONRPC)); err != nil {
+	// Use the actual request ID if available, otherwise fall back to expected ID for backward compatibility
+	var expectedID jsonrpc.ID
+	if requestID != nil && !requestID.IsEmpty() {
+		expectedID = *requestID
+	} else {
+		expectedID = getExpectedResponseID(jsonrpcResponse, isJSONRPC)
```

Comment:
Edge case to consider: if the JSONRPC request's ID is null, the response should match with a null in ID.